File: src/services/syncService.ts
import {
  initDatabase,
  getDatabase,
  getPendingRows,
  getDeletedRows,
  markSynced,
  applyServerRow,
  applyServerDelete,
  getSyncMeta,
  setSyncMeta,
  saveDatabase,
  SYNC_TABLES,
} from '../db/database.js';

export type SyncStatus = 'idle' | 'pushing' | 'pulling' | 'synced' | 'error' | 'offline';

export interface SyncState {
  status: SyncStatus;
  lastSyncAt: number;
  pendingCount: number;
  conflictCount: number;
  error: string | null;
  isOnline: boolean;
}

const API_BASE = () => getSyncMeta('api_url') || (import.meta.env.VITE_API_URL as string) || 'http://localhost:3001';

export function setApiUrl(url: string) {
  setSyncMeta('api_url', url);
}

export function getApiUrl(): string {
  return getSyncMeta('api_url') || (import.meta.env.VITE_API_URL as string) || 'http://localhost:3001';
}

let syncTimer: ReturnType<typeof setInterval> | null = null;
let syncStateListeners: Array<(state: SyncState) => void> = [];
let currentState: SyncState = {
  status: 'idle',
  lastSyncAt: 0,
  pendingCount: 0,
  conflictCount: 0,
  error: null,
  isOnline: navigator.onLine,
};

// --- Network monitoring ---
function setupNetworkMonitoring() {
  window.addEventListener('online', () => {
    currentState.isOnline = true;
    notifyListeners();
    // Auto-sync when coming back online
    performFullSync();
  });

  window.addEventListener('offline', () => {
    currentState.isOnline = false;
    currentState.status = 'offline';
    notifyListeners();
  });
}

// --- State management ---
function notifyListeners() {
  currentState.pendingCount = countPending();
  currentState.conflictCount = countConflicts();
  currentState.lastSyncAt = parseInt(getSyncMeta('last_sync')) || 0;
  syncStateListeners.forEach(fn => fn({ ...currentState }));
}

export function subscribeToSyncState(listener: (state: SyncState) => void): () => void {
  syncStateListeners.push(listener);
  listener({ ...currentState });
  return () => {
    syncStateListeners = syncStateListeners.filter(fn => fn !== listener);
  };
}

function countPending(): number {
  const db = getDatabase();
  let count = 0;
  for (const table of SYNC_TABLES) {
    try {
      const result = db.exec(`SELECT COUNT(*) as cnt FROM ${table} WHERE sync_status = 'pending'`);
      if (result.length > 0) count += Number(result[0].values[0][0]);
    } catch { /* table might not exist yet */ }
  }
  return count;
}

function countConflicts(): number {
  const db = getDatabase();
  let count = 0;
  for (const table of SYNC_TABLES) {
    try {
      const result = db.exec(`SELECT COUNT(*) as cnt FROM ${table} WHERE sync_status = 'conflict'`);
      if (result.length > 0) count += Number(result[0].values[0][0]);
    } catch { /* table might not exist yet */ }
  }
  return count;
}

// --- API helpers ---
async function apiRequest(path: string, options: RequestInit = {}): Promise<Response> {
  const token = getSyncMeta('auth_token');
  const headers: Record<string, string> = {
    'Content-Type': 'application/json',
    ...(token ? { Authorization: `Bearer ${token}` } : {}),
    ...((options.headers as Record<string, string>) || {}),
  };

  const response = await fetch(`${API_BASE()}${path}`, {
    ...options,
    headers,
  });

  if (!response.ok) {
    const body = await response.text();
    throw new Error(`API Error ${response.status}: ${body}`);
  }

  return response;
}

// --- Retry with exponential backoff ---
async function withRetry<T>(fn: () => Promise<T>, maxRetries = 3, baseDelay = 2000): Promise<T> {
  let lastError: Error | null = null;

  for (let attempt = 0; attempt < maxRetries; attempt++) {
    try {
      return await fn();
    } catch (err) {
      lastError = err as Error;
      if (attempt < maxRetries - 1) {
        const delay = baseDelay * Math.pow(2, attempt);
        console.warn(`Retry attempt ${attempt + 1}/${maxRetries} after ${delay}ms`, err);
        await new Promise(resolve => setTimeout(resolve, delay));
      }
    }
  }

  throw lastError!;
}

// --- Compression ---
function compressData(data: Record<string, any[]>): string {
  const json = JSON.stringify(data);
  try {
    return json; // In browser, we rely on HTTP compression (gzip)
    // For a real implementation, use pako or fflate for client-side compression
  } catch {
    return json;
  }
}

// --- PUSH ---
async function pushToServer(): Promise<{ pushed: number; conflicts: any[] }> {
  const deviceId = getSyncMeta('device_id') || 'unknown';
  const data: Record<string, any[]> = {};
  let totalPushed = 0;
  const allConflicts: any[] = [];

  // Gather pending rows
  for (const table of SYNC_TABLES) {
    try {
      const pending = getPendingRows(table);
      if (pending.length > 0) {
        data[table] = pending;
      }

      // Also gather soft-deleted rows that are pending
      const lastSync = parseInt(getSyncMeta('last_sync')) || 0;
      const deleted = getDeletedRows(table, lastSync);
      if (deleted.length > 0) {
        data[`_deleted_${table}`] = deleted;
      }
    } catch { /* skip */ }
  }

  if (Object.keys(data).length === 0) {
    return { pushed: 0, conflicts: [] };
  }

  const response = await withRetry(() =>
    apiRequest('/sync/push', {
      method: 'POST',
      body: JSON.stringify({ device_id: deviceId, data }),
    })
  );

  const result = await response.json();

  if (result.success) {
    // Mark pushed rows as synced
    for (const [tableName, info] of Object.entries(result.results || {}) as any[]) {
      if (info.pushed > 0 && data[tableName]) {
        const ids = data[tableName].map((row: any) => row.id);
        // Only mark non-deleted rows
        try {
          const db = getDatabase();
          const nonDeletedIds = data[tableName]
            .filter((row: any) => !row.deleted_at)
            .map((row: any) => row.id);
          markSynced(tableName, nonDeletedIds);
        } catch { /* skip */ }
      }

      if (info.conflicts && info.conflicts.length > 0) {
        allConflicts.push(...info.conflicts);
        // Mark conflicts locally
        for (const conflict of info.conflicts) {
          try {
            const db = getDatabase();
            db.run(`UPDATE ${tableName} SET sync_status = 'conflict' WHERE id = ?`, [conflict.id]);
          } catch { /* skip */ }
        }
      }
    }

    totalPushed = Object.values(result.results || {})
      .reduce((sum: number, info: any) => sum + (info.pushed || 0), 0) as number;

    saveDatabase();
  }

  return { pushed: totalPushed, conflicts: allConflicts };
}

// --- PULL ---
async function pullFromServer(): Promise<{ pulled: number }> {
  const lastSync = parseInt(getSyncMeta('last_sync')) || 0;

  const response = await withRetry(() =>
    apiRequest(`/sync/pull?last_sync=${lastSync}`)
  );

  const result = await response.json();
  const serverData = result.data || {};
  let pulledCount = 0;

  // Apply each table's data
  for (const [key, rows] of Object.entries(serverData) as any[]) {
    if (key.startsWith('_deleted_')) {
      // Handle deleted records
      const tableName = key.replace('_deleted_', '');
      for (const row of rows) {
        applyServerDelete(tableName, row.id, row.deleted_at, row.updated_at);
        pulledCount++;
      }
    } else {
      // Handle upsert records
      if (SYNC_TABLES.includes(key)) {
        for (const row of rows) {
          applyServerRow(key, row);
          pulledCount++;
        }
      }
    }
  }

  // Update last sync timestamp
  if (result.last_sync) {
    setSyncMeta('last_sync', String(result.last_sync));
  }

  saveDatabase();
  return { pulled: pulledCount };
}

// --- Full Sync ---
export async function performFullSync(): Promise<SyncState> {
  if (!currentState.isOnline) {
    currentState.status = 'offline';
    notifyListeners();
    return currentState;
  }

  const token = getSyncMeta('auth_token');
  if (!token) {
    currentState.status = 'idle';
    notifyListeners();
    return currentState;
  }

  try {
    // Step 1: Push local changes
    currentState.status = 'pushing';
    notifyListeners();

    const pushResult = await pushToServer();

    // Step 2: Pull server changes
    currentState.status = 'pulling';
    notifyListeners();

    const pullResult = await pullFromServer();

    // Done
    currentState.status = 'synced';
    currentState.error = null;
    notifyListeners();

    console.log(`✅ Sync complete: pushed ${pushResult.pushed}, pulled ${pullResult.pulled}, conflicts ${pushResult.conflicts.length}`);

    return currentState;
  } catch (err) {
    console.error('❌ Sync failed:', err);
    currentState.status = 'error';
    currentState.error = (err as Error).message;
    notifyListeners();
    return currentState;
  }
}

// --- Background Sync ---
export function startBackgroundSync(intervalMs = 30000): void {
  stopBackgroundSync();
  console.log(`🔄 Starting background sync every ${intervalMs / 1000}s`);

  syncTimer = setInterval(async () => {
    if (currentState.isOnline && getSyncMeta('auth_token')) {
      const pending = countPending();
      if (pending > 0 || currentState.status !== 'synced') {
        await performFullSync();
      }
    }
  }, intervalMs);

  // Also sync on visibility change (when user comes back to tab)
  document.addEventListener('visibilitychange', handleVisibilityChange);
}

export function stopBackgroundSync(): void {
  if (syncTimer) {
    clearInterval(syncTimer);
    syncTimer = null;
  }
  document.removeEventListener('visibilitychange', handleVisibilityChange);
}

function handleVisibilityChange() {
  if (document.visibilityState === 'visible' && currentState.isOnline) {
    performFullSync();
  }
}

// --- Initialize sync system ---
export function initSyncSystem(): void {
  setupNetworkMonitoring();
  notifyListeners();

  // Start background sync if logged in
  if (getSyncMeta('auth_token')) {
    startBackgroundSync();
  }
}

// --- Force resolve conflict (accept server or local) ---
export function resolveConflict(table: string, id: string, acceptServer: boolean): void {
  const db = getDatabase();

  if (acceptServer) {
    // Mark as synced - next pull will overwrite with server data
    db.run(`UPDATE ${table} SET sync_status = 'synced' WHERE id = ?`, [id]);
  } else {
    // Keep local - mark as pending so it gets pushed again
    db.run(`UPDATE ${table} SET sync_status = 'pending' WHERE id = ?`, [id]);
  }

  saveDatabase();
  notifyListeners();
}

// --- Auth helpers ---
export async function login(email: string, password: string): Promise<{ token: string; user: any }> {
  await initDatabase();
  const response = await apiRequest('/auth/login', {
    method: 'POST',
    body: JSON.stringify({ email, password }),
  });

  const result = await response.json();

  if (result.token) {
    setSyncMeta('auth_token', result.token);
    setSyncMeta('user_id', result.user.id);
    setSyncMeta('store_id', result.user.storeId);

    // Generate device ID if not set
    if (!getSyncMeta('device_id')) {
      setSyncMeta('device_id', crypto.randomUUID());
    }

    startBackgroundSync();
    await performFullSync();
  }

  return result;
}

export async function register(username: string, email: string, password: string, storeName: string): Promise<{ token: string; user: any }> {
  await initDatabase();
  const response = await apiRequest('/auth/register', {
    method: 'POST',
    body: JSON.stringify({ username, email, password, storeName }),
  });

  const result = await response.json();

  if (result.token) {
    setSyncMeta('auth_token', result.token);
    setSyncMeta('user_id', result.user.id);
    setSyncMeta('store_id', result.user.storeId);
    setSyncMeta('device_id', crypto.randomUUID());

    startBackgroundSync();
    await performFullSync();
  }

  return result;
}

export function logout(): void {
  stopBackgroundSync();
  setSyncMeta('auth_token', '');
  setSyncMeta('user_id', '');
  setSyncMeta('store_id', '');
  currentState.status = 'idle';
  notifyListeners();
}

File: mobile/lib/services/sync_service.dart
import 'dart:convert';
import 'dart:async';
import 'package:http/http.dart' as http;
import 'package:connectivity_plus/connectivity_plus.dart';
import '../db/app_database.dart';

enum SyncStatus { idle, pushing, pulling, synced, error, offline }

class SyncState {
  final SyncStatus status;
  final int lastSyncAt;
  final int pendingCount;
  final int conflictCount;
  final String? error;
  final bool isOnline;

  SyncState({
    this.status = SyncStatus.idle,
    this.lastSyncAt = 0,
    this.pendingCount = 0,
    this.conflictCount = 0,
    this.error,
    this.isOnline = true,
  });

  SyncState copyWith({
    SyncStatus? status,
    int? lastSyncAt,
    int? pendingCount,
    int? conflictCount,
    String? error,
    bool? isOnline,
  }) =>
      SyncState(
        status: status ?? this.status,
        lastSyncAt: lastSyncAt ?? this.lastSyncAt,
        pendingCount: pendingCount ?? this.pendingCount,
        conflictCount: conflictCount ?? this.conflictCount,
        error: error ?? this.error,
        isOnline: isOnline ?? this.isOnline,
      );
}

class SyncService {
  static String _apiUrl = 'http://localhost:3001';
  static String? _token;
  static SyncState _state = SyncState();
  static Timer? _syncTimer;
  static final _controller = StreamController<SyncState>.broadcast();
  static String? _deviceId;
  static const _syncTables = [
    'products',
    'categories',
    'orders',
    'order_items',
    'customers',
    'suppliers',
    'payments',
    'expenses',
    'employees',
    'prescriptions',
  ];

  static Stream<SyncState> get stateStream => _controller.stream;
  static SyncState get state => _state;

  static void setApiUrl(String url) {
    _apiUrl = url.replaceAll(RegExp(r'/+$'), '');
  }

  static String getApiUrl() => _apiUrl;

  static void _emit(SyncState s) {
    _state = s;
    _controller.add(s);
  }

  // ===== Auth =====

  static Future<void> login(String email, String password) async {
    final response = await _request('POST', '/auth/login', body: {
      'email': email,
      'password': password,
    });
    final data = jsonDecode(response.body);
    if (response.statusCode == 200) {
      _token = data['token'];
      await AppDatabase.saveAuthData(
        token: data['token'],
        userId: data['user']['id'],
        storeId: data['user']['store_id'] ?? '',
        username: data['user']['username'] ?? '',
      );
    } else {
      throw Exception(data['error'] ?? 'فشل تسجيل الدخول');
    }
  }

  static Future<void> register(
      String username, String email, String password, String storeName) async {
    final response = await _request('POST', '/auth/register', body: {
      'username': username,
      'email': email,
      'password': password,
      'store_name': storeName,
    });
    final data = jsonDecode(response.body);
    if (response.statusCode == 201) {
      _token = data['token'];
      await AppDatabase.saveAuthData(
        token: data['token'],
        userId: data['user']['id'],
        storeId: data['store_id'] ?? '',
        username: username,
      );
    } else {
      throw Exception(data['error'] ?? 'فشل إنشاء الحساب');
    }
  }

  static Future<void> logout() async {
    _token = null;
    await AppDatabase.clearAuthData();
  }

  static Future<bool> isLoggedIn() async {
    _token = await AppDatabase.authToken;
    return _token != null;
  }

  // ===== Sync System =====

  static Future<void> initSyncSystem() async {
    _token = await AppDatabase.authToken;

    _deviceId = await AppDatabase.getSyncMeta('device_id');
    if (_deviceId == null || _deviceId!.isEmpty) {
      _deviceId = AppDatabase.generateId();
      await AppDatabase.setSyncMeta('device_id', _deviceId!);
    }

    // Monitor connectivity
    Connectivity().onConnectivityChanged.listen((results) {
      final online = results.any((r) => r != ConnectivityResult.none);
      _emit(_state.copyWith(
          isOnline: online,
          status: online ? SyncStatus.idle : SyncStatus.offline));
      if (online) performFullSync();
    });

    // Background sync every 30s
    _syncTimer?.cancel();
    _syncTimer = Timer.periodic(const Duration(seconds: 30), (_) {
      if (_state.isOnline && _token != null) performFullSync();
    });
  }

  static Future<void> performFullSync() async {
    if (_token == null || !_state.isOnline) return;

    try {
      await _pushChanges();
      await _pullChanges();

      final pending = await _getTotalPending();
      final conflicts = await _getTotalConflicts();
      final lastSync = await AppDatabase.getSyncMeta('last_sync_at');

      _emit(_state.copyWith(
        status: SyncStatus.synced,
        pendingCount: pending,
        conflictCount: conflicts,
        lastSyncAt: int.tryParse(lastSync ?? '0') ?? 0,
        error: null,
      ));
    } catch (e) {
      _emit(_state.copyWith(status: SyncStatus.error, error: e.toString()));
    }
  }

  static Future<void> _pushChanges() async {
    _emit(_state.copyWith(status: SyncStatus.pushing));

    if (_deviceId == null || _deviceId!.isEmpty) {
      _deviceId = await AppDatabase.getSyncMeta('device_id');
      if (_deviceId == null || _deviceId!.isEmpty) {
        _deviceId = AppDatabase.generateId();
        await AppDatabase.setSyncMeta('device_id', _deviceId!);
      }
    }

    final Map<String, dynamic> data = {};

    for (final table in _syncTables) {
      final pending = await AppDatabase.getPending(table);
      if (pending.isEmpty) continue;

      data[table] = pending;
    }

    if (data.isEmpty) return;

    final response = await _request('POST', '/sync/push', body: {
      'device_id': _deviceId,
      'data': data,
    });

    if (response.statusCode == 200) {
      for (final entry in data.entries) {
        final table = entry.key;
        final rows = (entry.value as List).cast<Map<String, dynamic>>();
        for (final record in rows) {
          final id = record['id'];
          if (id is String && id.isNotEmpty) {
            await AppDatabase.markSynced(table, id);
          }
        }
      }
    }
  }

  static Future<void> _pullChanges() async {
    _emit(_state.copyWith(status: SyncStatus.pulling));

    final lastSync = await AppDatabase.getSyncMeta('last_sync') ??
        await AppDatabase.getSyncMeta('last_sync_at') ??
        '0';

    final response = await _request('GET', '/sync/pull?last_sync=$lastSync');
    if (response.statusCode != 200) return;

    final decoded = jsonDecode(response.body);
    final serverData = (decoded is Map && decoded['data'] is Map)
        ? Map<String, dynamic>.from(decoded['data'] as Map)
        : <String, dynamic>{};

    for (final entry in serverData.entries) {
      final key = entry.key;
      final rows = entry.value as List? ?? [];

      if (key.startsWith('_deleted_')) {
        final table = key.replaceFirst('_deleted_', '');
        for (final record in rows) {
          final map = Map<String, dynamic>.from(record as Map);
          final id = map['id'] as String?;
          final deletedAt = map['deleted_at'];
          final updatedAt = map['updated_at'];
          if (id == null) continue;
          if (deletedAt is int && updatedAt is int) {
            await AppDatabase.applyServerDelete(
                table, id, deletedAt, updatedAt);
          }
        }
        continue;
      }

      // Normal upserts
      for (final record in rows) {
        final map = Map<String, dynamic>.from(record as Map);
        final id = map['id'] as String?;
        if (id == null) continue;
        if (map['deleted_at'] != null) {
          final deletedAt = map['deleted_at'];
          final updatedAt = map['updated_at'];
          if (deletedAt is int && updatedAt is int) {
            await AppDatabase.applyServerDelete(key, id, deletedAt, updatedAt);
          }
        } else {
          await AppDatabase.applyServerData(key, map);
        }
      }
    }

    final lastSyncFromServer = decoded is Map ? decoded['last_sync'] : null;
    final newLastSync = (lastSyncFromServer is int)
        ? lastSyncFromServer
        : AppDatabase.timestamp();

    await AppDatabase.setSyncMeta('last_sync', newLastSync.toString());
    await AppDatabase.setSyncMeta('last_sync_at', newLastSync.toString());
  }

  static Future<int> _getTotalPending() async {
    int total = 0;
    for (final table in _syncTables) {
      total += await AppDatabase.getPendingCount(table);
    }
    return total;
  }

  static Future<int> _getTotalConflicts() async {
    if (AppDatabase.isWeb) return 0;
    final db = await AppDatabase.database;
    int total = 0;
    for (final table in _syncTables) {
      final result = await db!.rawQuery(
        "SELECT COUNT(*) as cnt FROM $table WHERE sync_status = 'conflict'",
      );
      total += result.first['cnt'] as int;
    }
    return total;
  }

  // ===== HTTP Helper with Retry =====

  static Future<http.Response> _request(
    String method,
    String path, {
    Map<String, dynamic>? body,
    int retries = 3,
  }) async {
    int attempt = 0;
    while (attempt < retries) {
      try {
        final uri = Uri.parse('$_apiUrl$path');
        final headers = <String, String>{
          'Content-Type': 'application/json',
        };
        if (_token != null) headers['Authorization'] = 'Bearer $_token';

        late http.Response response;
        switch (method) {
          case 'GET':
            response = await http.get(uri, headers: headers);
            break;
          case 'POST':
            response =
                await http.post(uri, headers: headers, body: jsonEncode(body));
            break;
          default:
            throw Exception('Unsupported method: $method');
        }
        return response;
      } catch (e) {
        attempt++;
        if (attempt >= retries) rethrow;
        await Future.delayed(Duration(seconds: attempt * 2));
      }
    }
    throw Exception('Max retries exceeded');
  }

  static void dispose() {
    _syncTimer?.cancel();
    _controller.close();
  }
}
